document/JS.bs

```diff
@@ -0,0 +1,935 @@
+    1. Let |module| be |moduleObject|.\[[Module]].
+    1. If |module|.imports is not an empty list, and |importObject| is undefined, throw a [=TypeError=] exception.
+    1. Let |funcs| be an empty [=list=] of callable JavaScript objects.
+    1. Let |memories| be an empty [=list=] of {{Memory}} objects.
+    1. Let |tables| be an empty [=list=] of {{Table}} objects.
+
+    Note: The |funcs|, |memories| and |tables| lists are collected and used in order to cache the JavaScript objects corresponding to WebAssembly objects. If a WebAssembly object is exported multiple times, it will appear in exports as the same object instance. Moreover, if a JavaScript wrapper of a WebAssembly object was imported, and then present on the export, a new wrapper will not need to be created. However, if an ordinary JS function is imported, it will appear as wrapped by an {{ExportedFunction}}.
+
+    6. Let |imports| be an empty [=list=] of [=external value=]s.
+    1. For each (|moduleName|, |componentName|, |externtype|) in [=module_imports=](|module|), do
+      1. Let |o| be ? [=Get=](|importObject|, |moduleName|).
+      1. If Type(|o|) is not [=Object=], throw a [=TypeError=] exception.
+      1. Let |v| be ? [=Get=](|o|, |componentName|)
+      1. If |externtype| is of the form 𝖿𝗎𝗇𝖼 |functype|,
+         1. If [=IsCallable=](|v|) is false, throw a {{LinkError}} exception.
+         1. If |v| is an {{ExportedFunction}},
+             1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
+
+             Note: The signature is checked by [=instantiate_module=] below.
+
+             2. Append |v| to |funcs|.
+         1. Otherwise,
+             1. Let |hostfunc| be a [=host function=] which performs the following steps when called:
+                 1. If the signature contains an [=i64=] (as argument or result), the host function immediately throws a {{TypeError}} when called.
+                 1. Let |arguments| be a [=list=] of the arguments of the invocation of this function.
+                 1. Let |jsArguments| be an empty [=list=].
+                 1. For each |arg| in |arguments|,
+                     1. Append [=ToJSValue=](|arg|) to |jsArguments|.
+                 1. Let |ret| be ? [=Call=](|v|, undefined, |jsArguments|). If an exception is thrown, trigger a WebAssembly trap, and propagate the exception to the enclosing JavaScript.
+                 1. Let [|parameters|] → [|results|] be |functype|.
+                 1. If |results| is empty, return undefined.
+                 1. Otherwise, return [=ToWebAssemblyValue=](|ret|, |results|[0]).
+             1. Let |store| be the current agent cluster's [=associated store=].
+             1. Let (|store|, |funcaddr|) be alloc_func(|store|, |functype|, |hostfunc|).
+             1. Set the current agent cluster's [=associated store=] to |store|.
```

Comment:
Can we link "current agent cluster"?

Reply:
I'm not sure what to link to for that. I don't see anything obvious in the JS spec. @syg @lars-t-hansen any thoughts?

Reply:
Based on pure pattern matching, I'd do the [surrounding agent](https://tc39.github.io/ecma262/#surrounding-agent)'s [agent cluster](https://tc39.github.io/ecma262/#sec-agent-clusters).